File: 0862-shortest-subarray-with-sum-at-least-k/0862-shortest-subarray-with-sum-at-least-k.java
class Solution {
    int binarySearch(int[] arr, List<int[]> list, int hi, int k) {
        int lo = 0;
        int ans = Integer.MAX_VALUE;
        while(lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            if(arr[0] - list.get(mid)[0] >= k) {
                ans = Math.min(ans, arr[1] - list.get(mid)[1]);
                lo = mid + 1;
            }
            else hi = mid - 1;
        }
        return ans;
    }
    public int shortestSubarray(int[] nums, int k) {
        List<int[]> list = new ArrayList<>();
        list.add(new int[]{0, 0});
        int sum = 0;
        int res = Integer.MAX_VALUE;
        for(int i=0; i<nums.length; i++) {
            sum += nums[i];
            if(sum < 0) sum = 0;
            if(sum >= list.get(list.size()-1)[0]) {
                list.add(new int[]{sum, i+1});
            }
            else {
                while(list.get(list.size()-1)[0] > sum) {
                    list.remove(list.size()-1);
                }
                list.add(new int[]{sum, i+1});
            }
            res = Math.min(res, binarySearch(list.get(list.size()-1), list, list.size()-2, k));
            if(res == 1) return res;
        }
        return res == Integer.MAX_VALUE ? -1 : res;
    }
}

File: 0862-shortest-subarray-with-sum-at-least-k/0862-shortest-subarray-with-sum-at-least-k.cpp
class Solution {
public:
    int shortestSubarray(vector<int>& nums, int k) {
        int n = nums.size();
        vector<long> prefix(n + 1);
        prefix[0] = 0;
        for(int i=0; i<n; i++) {
            prefix[i+1] = prefix[i] + (long)nums[i];
        }
        deque<int> dq;
        int ans = n + 1;
        for(int i=0; i<prefix.size(); i++) {
            while(!dq.empty() && prefix[i] <= prefix[dq.back()]) {
                dq.pop_back();
            }
            while(!dq.empty() && prefix[i] >= prefix[dq.front()] + k) {
                ans = min(ans, i - dq.front());
                dq.pop_front();
            }
            dq.push_back(i);
        }
        return ans == n + 1 ? -1 : ans;
    }
};